packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -914,17 +922,27 @@ class RenderParagraph extends RenderBox
             assert(false);
           }
         }
-        newChildren.add(
-          SemanticsNode()
-            ..updateWith(config: configuration)
-            ..rect = currentRect,
-        );
+        final SemanticsNode newChild = (_cachedChildNodes?.isNotEmpty == true)
+            ? _cachedChildNodes.removeFirst()
+            : SemanticsNode();
+        newChild
+          ..updateWith(config: configuration)
+          ..rect = currentRect;
+        newChildCache.addLast(newChild);
+        newChildren.add(newChild);
       }
       start += info.text.length;
     }
+    _cachedChildNodes = newChildCache;
```

Comment:
why don't we directly cache the reference of `newChildren` but create a new cache queue? It also seems weird we cache the entire semantics node so that we can reuse the id. Can we expose the method to set the id for the semantics node? If not, it seems like we should fix the castOS to deal with changing id?

Reply:
All RenderObject's cache their entire SemanticsNode in RenderObject._semantics to re-use ids. This method was kinda unusual because it didn't cache things.
The SemanticsNode takes over ownership of newChildren in SemanticsNodes.updateWith, so nobody outside of it should really be holding onto that list (nor modify it later).
I don't think it's unreasonable to expect stable IDs for SemanticsNodes that don't change. So castOS's expectation here seems fine.

Reply:
I see, I wasn't aware other renderobjects are caching the node.